needle/modules/dynamic/watch/traffic_capture.py

```diff
@@ -0,0 +1,163 @@
+            if chan in r:
+                data = chan.recv(1024)
+                if len(data) == 0:
+                    break
+                sock.send(data)
+        chan.close()
+        sock.close()
+
+    def _reverse_forward_tunnel(self, server_port, remote_host, remote_port, transport):
+        transport.request_port_forward('', server_port)
+        while True:
+            chan = transport.accept(1000)
+            if chan is None:
+                continue
+            # Directing traffic captured from server_port to remote_port on the
+            # workstation
+            self._handler(chan, remote_host, remote_port)
+
+    def _remote_portforward_start(self):
+
+        localhost = "127.0.0.1"
+
+        # Create new ssh connection
+        client = paramiko.SSHClient()
+        client.load_system_host_keys()
+
+        self.printer.debug('Connecting to ssh host %s:%d ...' %
+                           (self.device._ip, self.device._port))
+        try:
+            client.connect(self.device._ip, self.device._port,
+                           username=self.device._username, password=self.device._password)
+        except Exception as e:
+            self.printer.error('*** Failed to connect to %s:%d: %r' %
+                               (self.device._ip, self.device._port, e))
+            return
+
+        self.printer.debug('Now forwarding remote port %d to %s:%d ...' % (
+            self.options['device_port'], localhost, self.options['proxy_port']))
+
+        # Activate remote forwarding
+        self._reverse_forward_tunnel(self.options['device_port'], localhost, int(
+            self.options['proxy_port']), client.get_transport())
+
+    def _portforward_proxy_start(self):
+
+        self.tunnel = Process(target=self._remote_portforward_start)
+        self.tunnel.start()
+
+    def _portforward_proxy_stop(self):
+
+        self.tunnel.terminate()
+
+    # ==================================================================================================================
+    # RUN
+    # ==================================================================================================================
+    def module_run(self):
+        """Main Execution"""
+
+        # Uploading firewall rules
+        self.printer.info('Activating firewall rules...')
+        self.local_temp_file = self.local_op.build_temp_path_for_file(
+            "needle-pfctl.rules", self)
+        self.remote_temp_file = Constants.DEVICE_PATH_IFCTL_RULES
+        localhost = "127.0.0.1"
+
+        outbound_ports = self._parse_ports(self.options['outbound_ports'])
+        firewall_rules = 'rdr on lo0 inet proto tcp from any to any port {} -> {} port {}\npass out route-to (lo0 {}) inet proto tcp from any to any port {}\n'.format(
+            outbound_ports, localhost, self.options['device_port'], localhost, outbound_ports)
+        self.local_op.write_file(self.local_temp_file, firewall_rules)
+
+        self.device.remote_op.upload(
```

Comment:
you can write a file directly on the device rather than locally and then uploading it using 'remote_op.write_file(fname, body)`

For the complete list of APIs, see: https://github.com/mwrlabs/needle/wiki/Architecture